### PR TITLE
Support aggregation of just one of author/publish

### LINKED ISSUE #166: Support aggregation of just author or publish models with AemAggregator
The AemAggregator.aggregate() supports aggregation of user feature models with product feature models. However it always builds models for both author and publish. 
In some cases we only need a single one of these: author _or_ publish.

We should add support for the AemAggregator to support this too.

### PR DESCRIPTION
**Pull request by bosschaert**
This fixes #166

### CODE REVIEW

File: aemanalyser-core/src/main/java/com/adobe/aem/analyser/Runmode.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/*************************************************************************
+ * ADOBE CONFIDENTIAL
```
Comment (kwin) [inline aemanalyser-core/src/main/java/com/adobe/aem/analyser/Runmode.java:2]: Rather name this `ServiceType`, run mode is IMHO a broader concept and includes more dimensions than just author/publish.

File: aemanalyser-core/src/main/java/com/adobe/aem/analyser/Runmode.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * __________________
+ *
+ *  Copyright 2022 Adobe
+ *  All Rights Reserved.
+ *
+ * NOTICE: All information contained herein is, and remains
```
Comment (kwin) [inline aemanalyser-core/src/main/java/com/adobe/aem/analyser/Runmode.java:8]: The repo claims that this code is supposed to be licensed under ASL 2.0, this header does not indicate that.

Comment (bosschaert) [reply aemanalyser-core/src/main/java/com/adobe/aem/analyser/Runmode.java:8]: Whoops thanks, @kwin ! That's wrong. I'll fix that.

File: aemanalyser-core/src/main/java/com/adobe/aem/analyser/Runmode.java (hunk 1 above)
Comment (bosschaert) [reply aemanalyser-core/src/main/java/com/adobe/aem/analyser/Runmode.java]: I've made this change, thanks for the suggestion.

File: aemanalyser-core/src/main/java/com/adobe/aem/analyser/AemAggregator.java [hunk 3]
```diff
@@ -189,6 +191,20 @@ public void setFeatureOutputDirectory(final File featureOutputDirectory) {
         this.featureOutputDirectory = featureOutputDirectory;
     }
 
+    /**
+     * @return the runmodes
+     */
+    public ServiceType[] getRunmodes() {
```
Comment (cziegeler) [inline aemanalyser-core/src/main/java/com/adobe/aem/analyser/AemAggregator.java:197]: Nitpicking, but I think we usually use runMode (upper case M) :)

File: aemanalyser-core/src/main/java/com/adobe/aem/analyser/AemAnalyserUtil.java [hunk 4]
```diff
@@ -35,12 +39,33 @@ public class AemAnalyserUtil {
...
+    static final List<String> ALL_USED_MODES = Stream.concat(
+        AUTHOR_USED_MODES.stream(), PUBLISH_USED_MODES.stream()).collect(Collectors.toList());
 
     /** Default runmode */
     private static final String DEFAULT_MODE = "(default)";
 
+    static List<String> getUsedModes(ServiceType[] runmodes) {
+        Set<ServiceType> rm = new HashSet<>(Arrays.asList(runmodes));
+
+        if (rm.size() == 0) {
```
Comment (cziegeler) [inline aemanalyser-core/src/main/java/com/adobe/aem/analyser/AemAnalyserUtil.java:55]: This looks like an error condition to me. It seems the code expects that the arrray exists and has at least one value.
Thinking of it, maybe an EnumSet is the better type than an array, as it ensures that a value is in there only once

**Comment by cziegeler** [comment]
PR looks good to me, I would switch to an EnumSet instead of an array for the service types

**Review by cziegeler — APPROVED** [review]
lgtm, thanks

File: aemanalyser-core/src/main/java/com/adobe/aem/analyser/AemAnalyserUtil.java (hunk 4 above)
Comment (bosschaert) [reply aemanalyser-core/src/main/java/com/adobe/aem/analyser/AemAnalyserUtil.java]: I made it throw an IllegalStateException if the size is 0, hope that works for you.

File: aemanalyser-core/src/main/java/com/adobe/aem/analyser/AemAggregator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bosschaert) [reply aemanalyser-core/src/main/java/com/adobe/aem/analyser/AemAggregator.java]: Done

**Pull request merged**